zapgrpc/zapgrpc.go

```diff
@@ -36,65 +39,129 @@ func (f optionFunc) apply(log *Logger) {
 
 // WithDebug configures a Logger to print at zap's DebugLevel instead of
 // InfoLevel.
+// It only affects the Printf, Println and Print methods, which are only used in the gRPC v1 grpclog.Logger API.
+// Deprecated: use grpclog.SetLoggerV2() for v2 API.
 func WithDebug() Option {
 	return optionFunc(func(logger *Logger) {
-		logger.print = (*zap.SugaredLogger).Debug
-		logger.printf = (*zap.SugaredLogger).Debugf
+		logger.printToDebug = true
 	})
 }
 
 // NewLogger returns a new Logger.
-//
-// By default, Loggers print at zap's InfoLevel.
 func NewLogger(l *zap.Logger, options ...Option) *Logger {
 	logger := &Logger{
-		log:    l.Sugar(),
-		fatal:  (*zap.SugaredLogger).Fatal,
-		fatalf: (*zap.SugaredLogger).Fatalf,
-		print:  (*zap.SugaredLogger).Info,
-		printf: (*zap.SugaredLogger).Infof,
+		delegate:     l.Sugar(),
+		printToDebug: false,
+		fatalToWarn:  false,
 	}
 	for _, option := range options {
 		option.apply(logger)
 	}
 	return logger
 }
 
-// Logger adapts zap's Logger to be compatible with grpclog.Logger.
+// Logger adapts zap's Logger to be compatible with grpclog.LoggerV2 and the deprecated grpclog.Logger.
 type Logger struct {
-	log    *zap.SugaredLogger
-	fatal  func(*zap.SugaredLogger, ...interface{})
-	fatalf func(*zap.SugaredLogger, string, ...interface{})
-	print  func(*zap.SugaredLogger, ...interface{})
-	printf func(*zap.SugaredLogger, string, ...interface{})
-}
-
-// Fatal implements grpclog.Logger.
-func (l *Logger) Fatal(args ...interface{}) {
-	l.fatal(l.log, args...)
-}
-
-// Fatalf implements grpclog.Logger.
-func (l *Logger) Fatalf(format string, args ...interface{}) {
-	l.fatalf(l.log, format, args...)
-}
-
-// Fatalln implements grpclog.Logger.
-func (l *Logger) Fatalln(args ...interface{}) {
-	l.fatal(l.log, args...)
+	delegate     *zap.SugaredLogger
+	printToDebug bool
+	fatalToWarn  bool
 }
 
 // Print implements grpclog.Logger.
+// Deprecated: use Info().
 func (l *Logger) Print(args ...interface{}) {
-	l.print(l.log, args...)
+	if l.printToDebug {
+		l.delegate.Debug(args...)
+	} else {
+		l.delegate.Info(args...)
+	}
 }
 
 // Printf implements grpclog.Logger.
+// Deprecated: use Infof().
 func (l *Logger) Printf(format string, args ...interface{}) {
-	l.printf(l.log, format, args...)
+	if l.printToDebug {
+		l.delegate.Debugf(format, args...)
+	} else {
+		l.delegate.Infof(format, args...)
+	}
 }
 
 // Println implements grpclog.Logger.
+// Deprecated: use Info().
 func (l *Logger) Println(args ...interface{}) {
-	l.print(l.log, args...)
+	l.Print(args...)
+}
+
+// Info implements grpclog.LoggerV2.
+func (l *Logger) Info(args ...interface{}) {
+	l.delegate.Info(args...)
+}
+
+// Infoln implements grpclog.LoggerV2.
+func (l *Logger) Infoln(args ...interface{}) {
```

Comment:
The *Ln methods are slightly different from the non-Ln methods, the Ln methods always add spaces between all arguments, while the non-Ln methods only add spaces between objects if the objects are not both strings,
https://golang.org/pkg/fmt/#Println

See related issue on zap which doesn't have *Ln methods,
https://github.com/uber-go/zap/issues/889

I think the existing `Println` method is similarly incorrect, just didn't catch this till now

Reply:
Hm, I was not aware of such nuance. Done.